packages/controls/src/controls/style/style.ts

```diff
@@ -0,0 +1,237 @@
+> {
+  static readonly type = 'makeswift::controls::style' as const
+  static readonly Property = {
+    Width: `${this.type}::property::width`,
+    Margin: `${this.type}::property::margin`,
+    Padding: `${this.type}::property::padding`,
+    Border: `${this.type}::property::border`,
+    BorderRadius: `${this.type}::property::border-radius`,
+    TextStyle: `${this.type}::property::text-style`,
+  } as const
+
+  static readonly DefaultProperties: StyleProperty[] = [
+    this.Property.Width,
+    this.Property.Margin,
+  ]
+
+  static readonly AllProperties: StyleProperty[] = Object.values(this.Property)
+
+  static get schema() {
+    const type = z.literal(Definition.type)
+
+    const properties = z.union([
+      z.literal(this.Property.Width),
+      z.literal(this.Property.Margin),
+      z.literal(this.Property.Padding),
+      z.literal(this.Property.Border),
+      z.literal(this.Property.BorderRadius),
+      z.literal(this.Property.TextStyle),
+    ])
+
+    const config = z.object({
+      properties: z.array(properties),
+    })
+
+    const dataSchema = <C extends z.ZodTypeAny>(color: C) =>
+      z.object({
+        width: responsiveValue(StyleSchema.width).optional(),
+        margin: responsiveValue(StyleSchema.margin).optional(),
+        padding: responsiveValue(StyleSchema.padding).optional(),
+        border: responsiveValue(StyleSchema.borderSchema(color)).optional(),
+        borderRadius: responsiveValue(StyleSchema.borderRadius).optional(),
+        textStyle: responsiveValue(StyleSchema.textStyle).optional(),
+      })
+
+    const data = dataSchema(Schema.colorData)
+    const resolvedData = dataSchema(Schema.resolvedColorData)
+
+    const value = data
+    const resolvedValue = z.string()
+
+    const definition = z.object({
+      type,
+      config,
+    })
+
+    return {
+      type,
+      config,
+      definition,
+      data,
+      resolvedData,
+      value,
+      resolvedValue,
+    }
+  }
+
+  static deserialize(data: DeserializedRecord): StyleDefinition {
+    if (data.type !== Definition.type) {
+      throw new Error(
+        `Style: expected type ${Definition.type}, got ${data.type}`,
+      )
+    }
+
+    const { config } = Definition.schema.definition.parse(data)
+    return Style(config)
+  }
+
+  get controlType() {
+    return Definition.type
+  }
+
+  get schema() {
+    return {
+      ...Definition.schema,
+      data: Definition.schema.data as SchemaType<DataType<C>>,
+      value: Definition.schema.value as SchemaType<ValueType<C>>,
+    }
+  }
+
+  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
+    return safeParse(this.schema.data, data)
+  }
+
+  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
+    return this.schema.data.optional().parse(data)
+  }
+
+  toData(value: ValueType<C>): DataType<C> {
+    return value
+  }
+
+  copyData(
+    data: DataType<C> | undefined,
+    context: CopyContext,
+  ): DataType<C> | undefined {
```

Comment:
Tidied up version of https://github.com/makeswift/makeswift/pull/781/files#diff-fe2757ce8e16ddd867978a37261e04fe23962f06fdaafd490fc0b26f226411ceL244